### PR TITLE
[IMP] project,{hr,sale}_timesheet: improve portal search view

### PR DESCRIPTION
**Pull request by MissingNoShiny**
This commit aims to bring the portal search view closer to that of the back end.
In order to do that, it adds, removes and reorganizes some search fields, filters and groupbys.

Enterprise: https://github.com/odoo/enterprise/pull/46173

Task-3444823

### CODE REVIEW

File: addons/hr_timesheet/views/hr_timesheet_portal_templates.xml [hunk 1]
```diff
@@ -100,6 +100,47 @@
...
+                                        <span t-else="">No Parent Task</span>
+                                    </th>
+                                    <th colspan="1" class="text-end text-muted">
+                                        <t t-if="is_uom_day">
+                                            Total: <span t-esc="timesheets._convert_hours_to_days(hours_spent)" t-options='{"widget": "timesheet_uom"}'/>
+                                        </t>
+                                        <t t-else="">
+                                            Total: <span t-esc="hours_spent" t-options='{"widget": "float_time"}'/>
+                                        </t>
+                                    </th>
```
Comment (pkyriakou) [inline addons/hr_timesheet/views/hr_timesheet_portal_templates.xml:115]: Since this is repeated three time below, I am wondering whether we can do the t-if and only change the top part for each case (an example top part):
```xml
<th colspan="6">
    <span t-if="timesheets[0].parent_task_id" t-field="timesheets[0].parent_task_id.name"/>
    <span t-else="">No Parent Task</span>
</th>
```

**Review by pkyriakou — COMMENTED** [review]
Thank you @MissingNoShiny 
I left one comment :100:

File: addons/hr_timesheet/views/hr_timesheet_portal_templates.xml (hunk 1 above)
Comment (MissingNoShiny) [reply addons/hr_timesheet/views/hr_timesheet_portal_templates.xml]: I originally meant to do that bu I guess I forgot, done now :+1:

**Review by pkyriakou — COMMENTED** [review]
Thank you for the changes, the look extra good to me :partying_face:

**Review by pkyriakou — COMMENTED** [review]
Thank you @MissingNoShiny for the updates :100: 
Left you two questions/comments and then we good to go :partying_face:

File: addons/sale_project/controllers/portal.py [hunk 2]
```diff
@@ -12,30 +12,29 @@ class SaleProjectCustomerPortal(ProjectCustomerPortal):
...
 
     def _task_get_search_domain(self, search_in, search):
         search_domain = [super()._task_get_search_domain(search_in, search)]
-        if search_in in ('sale_order', 'all'):
+        if search_in == 'sale_order':
             search_domain.append(['|', ('sale_order_id.name', 'ilike', search), ('sale_line_id.name', 'ilike', search)])
-        if search_in in ('invoice', 'all'):
+        if search_in == 'invoice':
             search_domain.append([('sale_order_id.invoice_ids.name', 'ilike', search)])
         return OR(search_domain)
```
[the remark this answers is not in the document]
Comment (MissingNoShiny) [reply addons/sale_project/controllers/portal.py:40]: Indeed, I missed it, thanks :+1:

File: addons/sale_timesheet/controllers/portal.py [hunk 3]
```diff
@@ -60,46 +60,30 @@ def portal_my_tasks_invoices(self, task_id=None, page=1, date_begin=None, date_e
...
     def _get_searchbar_inputs(self):
-        searchbar_inputs = super()._get_searchbar_inputs()
-        searchbar_inputs.update(
-            so={'input': 'so', 'label': _('Search in Sales Order')},
-            sol={'input': 'sol', 'label': _('Search in Sales Order Item')},
-            invoice={'input': 'invoice', 'label': _('Search in Invoice')})
-        return searchbar_inputs
+        return super()._get_searchbar_inputs() | {
+            'so': {'input': 'so', 'label': _('Search in Sales Order'), 'sequence': 40},
+            'invoice': {'input': 'invoice', 'label': _('Search in Invoice'), 'sequence': 80},
```
[the remark this answers is not in the document]
Comment (MissingNoShiny) [reply addons/sale_timesheet/controllers/portal.py:65]: I discussed with Louis, who works on translations, and apparently, `_lt` should only be used for terms that are top level in the python file, i.e. outside of any method. I changed the usages of `_lt` to `_`

**Review by pkyriakou — COMMENTED** [review]
Alright great :100: thank you for the changes @MissingNoShiny 
LGTM :green_circle: :partying_face:

File: addons/project/controllers/portal.py [hunk 4]
```diff
@@ -313,97 +313,64 @@ def _task_get_page_view_values(self, task, access_token, **kwargs):
...
-            'all': {'input': 'all', 'label': _('Search in All'), 'order': 1},
-            'content': {'input': 'content', 'label': Markup(_('Search <span class="nolabel"> (in Content)</span>')), 'order': 1},
-            'ref': {'input': 'ref', 'label': _('Search in Ref'), 'order': 1},
-            'users': {'input': 'users', 'label': _('Search in Assignees'), 'order': 3},
-            'stage': {'input': 'stage', 'label': _('Search in Stages'), 'order': 4},
-            'status': {'input': 'status', 'label': _('Search in Status'), 'order': 5},
-            'priority': {'input': 'priority', 'label': _('Search in Priority'), 'order': 7},
-            'customer': {'input': 'customer', 'label': _('Search in Customer'), 'order': 10},
-            'message': {'input': 'message', 'label': _('Search in Messages'), 'order': 11},
+            'name': {'input': 'name', 'label': Markup(_('Search<span class="nolabel"> Tasks</span>')), 'sequence': 10},
```
Comment (xavierbol) [inline addons/project/controllers/portal.py:347]: Hello @odoo/rd-security 

Can we have an override?

We just move the Markup function. 🙂 

Thanks in advance!

Have a nice afternoon! 🙂

File: addons/sale_project/controllers/portal.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import _lt
+from odoo import _
```
Comment (xavierbol) [inline addons/sale_project/controllers/portal.py:4]: Why did you change the function used for the translation? 🧐

File: addons/project/controllers/portal.py [hunk 6]
```diff
@@ -313,97 +313,64 @@ def _task_get_page_view_values(self, task, access_token, **kwargs):
...
+            'users': {'input': 'user_ids', 'label': _('Search in Assignees'), 'sequence': 20},
+            'stage_id': {'input': 'stage_id', 'label': _('Search in Stages'), 'sequence': 30},
+            'status': {'input': 'status', 'label': _('Search in Status'), 'sequence': 40},
+            'priority': {'input': 'priority', 'label': _('Search in Priority'), 'sequence': 60},
+            'partner_id': {'input': 'partner_id', 'label': _('Search in Customer'), 'sequence': 80},
```
Comment (xavierbol) [inline addons/project/controllers/portal.py:352]: I am wondering if it should not be better to use `_lt` instead of `_` to translate those terms at the end. 🤔

File: addons/project/controllers/portal.py [hunk 7]
```diff
@@ -313,97 +313,64 @@ def _task_get_page_view_values(self, task, access_token, **kwargs):
...
         }
         if not project:
-            values['project'] = {'input': 'project', 'label': _('Search in Project'), 'order': 2}
+            values['project_id'] = {'input': 'project_id', 'label': _('Search in Project'), 'sequence': 50}
         if milestones_allowed:
-            values['milestone'] = {'input': 'milestone', 'label': _('Search in Milestone'), 'order': 6}
+            values['milestone_id'] = {'input': 'milestone_id', 'label': _('Search in Milestone'), 'sequence': 70}
```
Comment (xavierbol) [inline addons/project/controllers/portal.py:357]: same comment here. 🙂

File: addons/sale_project/controllers/portal.py [hunk 8]
```diff
@@ -12,30 +12,28 @@ class SaleProjectCustomerPortal(ProjectCustomerPortal):
...
-            search_domain.append(['|', ('sale_order_id.name', 'ilike', search), ('sale_line_id.name', 'ilike', search)])
-        if search_in in ('invoice', 'all'):
-            search_domain.append([('sale_order_id.invoice_ids.name', 'ilike', search)])
-        return OR(search_domain)
+        if search_in == 'sale_order':
+            return ['|', ('sale_order_id.name', 'ilike', search), ('sale_line_id.name', 'ilike', search)]
+        elif search_in == 'invoice':
+            return [('sale_order_id.invoice_ids.name', 'ilike', search)]
+        else:
+            return super()._task_get_search_domain(search_in, search)
```
Comment (xavierbol) [inline addons/sale_project/controllers/portal.py:39]: Equivalent to:
```suggestion
        return super()._task_get_search_domain(search_in, search)
```

File: addons/hr_timesheet/controllers/portal.py [hunk 9]
```diff
@@ -26,11 +25,11 @@ def _prepare_home_portal_values(self, counters):
...
-            'all': {'input': 'all', 'label': _('Search in All')},
-            'employee': {'input': 'employee', 'label': _('Search in Employee')},
-            'project': {'input': 'project', 'label': _('Search in Project')},
-            'task': {'input': 'task', 'label': _('Search in Task')},
-            'name': {'input': 'name', 'label': _('Search in Description')},
+            'name': {'input': 'name', 'label': _('Search in Description'), 'sequence': 10},
+            'employee_id': {'input': 'employee_id', 'label': _('Search in Employee'), 'sequence': 20},
+            'project_id': {'input': 'project_id', 'label': _('Search in Project'), 'sequence': 30},
+            'task_id': {'input': 'task_id', 'label': _('Search in Task'), 'sequence': 40},
+            'parent_task_id': {'input': 'parent_task_id', 'label': _('Search in Parent Task'), 'sequence': 70},
```
Comment (xavierbol) [inline addons/hr_timesheet/controllers/portal.py:32]: I am wondering if it should not be better to use `_lt` function instead of `_` function. 🤔

File: addons/hr_timesheet/controllers/portal.py [hunk 10]
```diff
@@ -40,42 +39,24 @@ def _task_get_searchbar_sortings(self, milestones_allowed, project=False):
...
-            'project': {'input': 'project', 'label': _('Project')},
-            'task': {'input': 'task', 'label': _('Task')},
-            'date': {'input': 'date', 'label': _('Date')},
-            'employee': {'input': 'employee', 'label': _('Employee')}
+            'none': {'label': _('None'), 'sequence': 10},
+            'date': {'label': _('Date'), 'sequence': 20},
+            'project_id': {'label': _('Project'), 'sequence': 30},
+            'parent_task_id': {'label': _('Parent Task'), 'sequence': 40},
+            'task_id': {'label': _('Task'), 'sequence': 50},
+            'employee_id': {'label': _('Employee'), 'sequence': 70},
```
Comment (xavierbol) [inline addons/hr_timesheet/controllers/portal.py:47]: Same comment here

File: addons/hr_timesheet/controllers/portal.py [hunk 11]
```diff
@@ -40,42 +39,24 @@ def _task_get_searchbar_sortings(self, milestones_allowed, project=False):
...
-            'date': {'label': _('Newest'), 'order': 'date desc'},
-            'employee': {'label': _('Employee'), 'order': 'employee_id'},
-            'project': {'label': _('Project'), 'order': 'project_id'},
-            'task': {'label': _('Task'), 'order': 'task_id'},
-            'name': {'label': _('Description'), 'order': 'name'},
+            'date desc': {'label': _('Newest')},
+            'employee_id': {'label': _('Employee')},
+            'project_id': {'label': _('Project')},
+            'task_id': {'label': _('Task')},
+            'name': {'label': _('Description')},
```
Comment (xavierbol) [inline addons/hr_timesheet/controllers/portal.py:59]: Same comment here. 🙂

File: addons/sale_timesheet/controllers/portal.py [hunk 12]
```diff
@@ -60,46 +60,30 @@ def portal_my_tasks_invoices(self, task_id=None, page=1, date_begin=None, date_e
...
     def _get_searchbar_inputs(self):
-        searchbar_inputs = super()._get_searchbar_inputs()
-        searchbar_inputs.update(
-            so={'input': 'so', 'label': _('Search in Sales Order')},
-            sol={'input': 'sol', 'label': _('Search in Sales Order Item')},
-            invoice={'input': 'invoice', 'label': _('Search in Invoice')})
-        return searchbar_inputs
+        return super()._get_searchbar_inputs() | {
+            'so': {'input': 'so', 'label': _('Search in Sales Order'), 'sequence': 50},
+            'invoice': {'input': 'invoice', 'label': _('Search in Invoice'), 'sequence': 80},
```
Comment (xavierbol) [inline addons/sale_timesheet/controllers/portal.py:65]: Same comment here. 🙂

File: addons/sale_timesheet/controllers/portal.py [hunk 13]
```diff
@@ -60,46 +60,30 @@ def portal_my_tasks_invoices(self, task_id=None, page=1, date_begin=None, date_e
...
     def _get_searchbar_groupby(self):
-        searchbar_groupby = super()._get_searchbar_groupby()
-        searchbar_groupby.update(
-            so={'input': 'so', 'label': _('Sales Order')},
-            sol={'input': 'sol', 'label': _('Sales Order Item')},
-            invoice={'input': 'invoice', 'label': _('Invoice')})
-        return searchbar_groupby
+        return super()._get_searchbar_groupby() | {
+            'so_line': {'label': _('Sales Order Item'), 'sequence': 80},
+            'timesheet_invoice_id': {'label': _('Invoice'), 'sequence': 90},
```
Comment (xavierbol) [inline addons/sale_timesheet/controllers/portal.py:71]: Same comment here. 🙂

File: addons/sale_timesheet/controllers/portal.py [hunk 14]
```diff
@@ -60,46 +60,30 @@ def portal_my_tasks_invoices(self, task_id=None, page=1, date_begin=None, date_e
...
+        else:
+            return super()._get_search_domain(search_in, search)
 
     def _get_searchbar_sortings(self):
-        searchbar_sortings = super()._get_searchbar_sortings()
-        searchbar_sortings.update(
-            sol={'label': _('Sales Order Item'), 'order': 'so_line'})
-        return searchbar_sortings
+        return super()._get_searchbar_sortings() | {
+            'so_line': {'label': _('Sales Order Item')}
```
Comment (xavierbol) [inline addons/sale_timesheet/controllers/portal.py:85]: Same comment here. 🙂

**Review by xavierbol — COMMENTED** [review]
Thanks for your work, I left some comments. 🙂

File: addons/project/controllers/portal.py (hunk 4 above)
Comment (mart-e) [reply addons/project/controllers/portal.py:347]: Why is there that weird span on one word? if you rearrange you could do something nicer like
```suggestion
            'name': {'input': 'name', 'label': Markup('<span class="nolabel">%s</span>') % _('Search Tasks'), 'sequence': 10},
```
We try to avoid having translated html

Comment (mart-e) [reply addons/project/controllers/portal.py:347]: If you really want to have this span in the middle would be something like
```py
_("Search %(left)s Tasks %(right)s",
    left=Markup('<span class="nolabel">'),
    right=Markup('</span>'))
```

Comment (xavierbol) [reply addons/project/controllers/portal.py:347]: @mart-e hum indeed, I will let @MissingNoShiny check that.

Thanks for your review. 😘 

Have a nice afternoon!

File: addons/sale_project/controllers/portal.py (hunk 5 above)
Comment (MissingNoShiny) [reply addons/sale_project/controllers/portal.py:4]: I replied on the enterprise PR

File: addons/sale_project/controllers/portal.py (hunk 8 above)
Comment (MissingNoShiny) [reply addons/sale_project/controllers/portal.py:39]: I replied on the enterprise PR

File: addons/project/controllers/portal.py [hunk 15]
```diff
@@ -313,97 +313,68 @@ def _task_get_page_view_values(self, task, access_token, **kwargs):
...
-            'stage': {'input': 'stage', 'label': _('Search in Stages'), 'order': 4},
-            'status': {'input': 'status', 'label': _('Search in Status'), 'order': 5},
-            'priority': {'input': 'priority', 'label': _('Search in Priority'), 'order': 7},
-            'customer': {'input': 'customer', 'label': _('Search in Customer'), 'order': 10},
-            'message': {'input': 'message', 'label': _('Search in Messages'), 'order': 11},
+            'name': {'input': 'name', 'label': _(
+                'Search%(left)s Tasks%(right)s',
+                left=Markup('<span class="nolabel">'),
+                right=Markup('</span>'),
+            ), 'sequence': 10},
```
Comment (xavierbol) [inline addons/project/controllers/portal.py:351]: Thanks again for your review @mart-e 

@MissingNoShiny did the changes, he also did the same changes in the enterprise PR. 🙂 
https://github.com/odoo/enterprise/pull/46173

Can we have the override if it is good for you? 

Thanks in advance! 

Have a nice afternoon! 🙂

**Review by xavierbol — COMMENTED** [review]
LGTM, thanks for your work! 🙂 

robodoo r+

**Comment by mart-e** [comment]
> Can we have the override if it is good for you?

Not clear if we need it with the minimal CI but looks good now
@robodoo override=ci/security

**Comment by xavierbol** [comment]
robodoo r+

**Comment by xavierbol** [comment]
robodoo r+

**Comment by xavierbol** [comment]
robodoo rebase-ff
